### ISSUE TITLE
[Advanced Scripting] Create Multiple Partitions after a template

### ISSUE THREAD

**Issue by Rasenkantenstein** [otykier/TabularEditor#366]
Hi Daniel & folks,

I was again wondering whether someone could help me with creating partitions based on a template. The template partition would contain the string such as "template", i.e. FactSales_Template
In my case the partition is based on a month with a range of datekeys. So in the template I would add datekeys such as 
`datekey between 19000101 and 19001231`

The added partitions would be called something like "FactSales_201911" for november etc.

Thank you very much :-)
Konstantin

**Comment by Rasenkantenstein** [comment]
BTW:

I have found a script template on the internet, but I am not sure I fully understand what is going on:

`string SSASDataBaseName = null;
            string SSAS_TemplateMarker = null;// = Dts.Variables["SSAS_TemplateMarker"].Value.ToString();
            if (SSAS_TemplateMarker == null)
            {
                SSAS_TemplateMarker = "Template";
            }
  
  foreach (Table tab in Selected.Tables)
            {

                int containTemplate = 0;
                string MeasGrName = tab.Name;

                string partitionsToProcess = ""
                string PartitionNameTemplate = "";
                string SSAS_PartitionTemplateSourceSQL = "";
                // Find the template partition
                foreach (Partition part in tab.Partitions)
                {
                    int findTemplateStringFirst = part.Name.IndexOf(SSAS_TemplateMarker);
                    if (findTemplateStringFirst > 0)
                    {
                        containTemplate = 1;
                        PartitionNameTemplate = part.Name.Replace(SSAS_TemplateMarker, "");


                        SSAS_PartitionTemplateSourceSQL = part.Query;
                        SSAS_PartitionTemplateSourceSQL = SSAS_PartitionTemplateSourceSQL.Replace("AND 1 = 0", "");
                        SSAS_PartitionTemplateSourceSQL = SSAS_PartitionTemplateSourceSQL.Replace("AND 1=0", "");
                    }
                }
                //----------------
                if (containTemplate > 0)
                {

                    foreach (var partitionSuffix in partitionsToProcess.Split(','))
                    {

                        DateTime Month2Partition = DateTime.ParseExact(partitionSuffix + "01", "yyyyMMdd", System.Globalization.CultureInfo.InvariantCulture);
                        string SSAS_PartitionSourceSQL;
                        string PartitionName = PartitionNameTemplate + Month2Partition.ToString("yyyyMM");
                        string dateFrom = Month2Partition.ToString("yyyyMM") + "01";
                        string dateTo = Month2Partition.ToString("yyyyMM")
                            + (new DateTime(Month2Partition.Year, Month2Partition.Month, 1).AddMonths(1).AddDays(-1)).Day.ToString();

                        SSAS_PartitionSourceSQL = SSAS_PartitionTemplateSourceSQL.Replace("19000101", dateFrom).Replace("19001231", dateTo);
                        
						
                        //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
                        //-------------------------------------------- CREATE NEW PARTITION------------------------------------------------
                        // define the new partition
                        if ((tab.Partitions.Find(PartitionName) == null) && !DoNotCreatePartitionsFlag)
                        {
                            QueryPartitionSource src = new QueryPartitionSource();
                            src.Query = SSAS_PartitionSourceSQL;
                            src.DataSource = ds;
                            Partition p = new Partition();
                            p.Name = PartitionName;
                            p.Source = src;
                            p.Mode = ModeType.Default;
                            p.DataView = DataViewType.Default;
                            tab.Partitions.Add(p);
                                                    }
                        //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
                    } // loop Partitions
                }
            } // foreach`

**Comment by otykier** [comment]
As of Tabular Editor 2.8.6, it is actually not possible to create partitions through a script. This is an oversight on my part, but the good news is that it has already been implemented in 2.8.7, which will be released in a few days:

![image](https://user-images.githubusercontent.com/8976200/70134559-a5b9c500-1688-11ea-9329-d0567261540b.png)

Once released, 2.8.7 will let you add partitions through a script. To make this work, I suggest you define an annotation on the table, to hold the template SQL used by all partitions:

![image](https://user-images.githubusercontent.com/8976200/70135273-07c6fa00-168a-11ea-84f6-90f0b3498ed8.png)

The following script will then create all monthly partitions (note - it won't delete any pre-existing partitions on the table), based on the template SQL as well as the specified first- and last partition dates.

```csharp
var firstPartition = new DateTime(2018,1,1); // First partition date
var lastPartition = new DateTime(2020,12,1); // Last partition date

var templateSql = Selected.Table.GetAnnotation("PartitionTemplateSQL");
if(string.IsNullOrEmpty(templateSql)) throw new Exception("No partition template!");

var currentPartition = firstPartition;
while(currentPartition <= lastPartition)
{
    var calendarIdFrom = currentPartition.ToString("yyyyMMdd");
    var calendarIdTo = currentPartition.AddMonths(1).AddDays(-1).ToString("yyyyMMdd");
    var partitionName = Selected.Table.Name + "_" + currentPartition.ToString("yyyyMM");
    var partitionQuery = string.Format(templateSql, calendarIdFrom, calendarIdTo);
    Selected.Table.AddPartition(partitionName, partitionQuery);
    
    // Increment to next month:
    currentPartition = currentPartition.AddMonths(1);
}
```

**Comment by otykier** [comment]
[2.9.0](https://github.com/otykier/TabularEditor/releases/tag/2.9.0) has just been released, which lets you use the script shown above.
